cpp/src/io/shp/polygon_shp_reader.cu

```diff
@@ -0,0 +1,243 @@
+            
+            std::vector<double> aPointX;
+            std::vector<double> aPointY;
+            std::vector<int> aPartSize;
+ 	    PolygonFromGeometry( poShape, aPointX, aPointY, aPartSize );
+ 		
+            x_v.insert(x_v.end(),aPointX.begin(),aPointX.end());
+ 	    y_v.insert(y_v.end(),	aPointY.begin(),aPointY.end());
+            r_len_v.insert(r_len_v.end(),aPartSize.begin(),aPartSize.end());
+ 	    f_len_v.push_back(aPartSize.size());
+ 	    OGR_F_Destroy( hFeat );
+ 	    num_feature++;
+        }
+        g_len_v.push_back(num_feature);
+        return num_feature;
+    }
+
+    /*
+    * Read a polygon shapefile and fill in a polygons structure
+    * ToDo: read associated relational data into a CUDF Table 
+    *
+    * filename: ESRI shapefile name (wtih .shp extension
+    * pm: structure polygons (fixed to double type) to hold polygon data
+    
+    * Note: only the first layer is read - shapefiles have only one layer in GDALDataset model    
+    */
+
+    void read_shapefile(const char *filename, struct polygons<double>& pm)
+    {
+        vector<int> g_len_v,f_len_v,r_len_v;
+        vector<double> x_v, y_v;
+        GDALAllRegister();
+        
+        GDALDatasetH hDS = GDALOpenEx( filename, GDAL_OF_VECTOR, NULL, NULL, NULL );
+        CUDF_EXPECTS(hDS!=NULL,"Failed to open ESRI Shapefile dataset");		    
+        OGRLayerH hLayer = GDALDatasetGetLayer( hDS,0 );
+        CUDF_EXPECTS(hLayer!=NULL,"Failed to open the first layer");
+        int num_f=ReadLayer(hLayer,g_len_v,f_len_v,r_len_v,x_v,y_v);
+        CUDF_EXPECTS(num_f>0,"Shapefile must have at lest one polygon");
+        
+        pm.num_group=g_len_v.size();
+        pm.num_feature=f_len_v.size();
+        pm.num_ring=r_len_v.size();
+        pm.num_vertex=x_v.size();
+        pm.group_length=new uint32_t[ pm.num_group];
+        pm.feature_length=new uint32_t[ pm.num_feature];
+        pm.ring_length=new uint32_t[ pm.num_ring];
+        pm.x=new double [pm.num_vertex];
+        pm.y=new double [pm.num_vertex];
+        CUDF_EXPECTS(pm.group_length!=nullptr&&pm.feature_length!=nullptr&&pm.ring_length!=nullptr,"expecting p_{g,f,r}_len are non-zeron");		    
+        CUDF_EXPECTS(pm.x!=nullptr&&pm.y!=nullptr,"expecting polygon x/y arrays are not nullptr");
+	      
+        memcpy((void *)(pm.group_length),(void *)(g_len_v.data()),pm.num_group*sizeof(uint32_t));
+        memcpy((void *)(pm.feature_length),(void *)(f_len_v.data()),pm.num_feature*sizeof(uint32_t));
+        memcpy((void *)(pm.ring_length),(void *)(r_len_v.data()),pm.num_ring*sizeof(uint32_t));
+        memcpy((void *)(pm.x),(void *)(x_v.data()),pm.num_vertex*sizeof(double));
+        memcpy((void *)(pm.y),(void *)(y_v.data()),pm.num_vertex*sizeof(double));      
+    }
+ 
+    /*
+    * read polygon data from file in ESRI Shapefile format; data type of vertices is fixed to double (GDF_FLOAT64)
+    * see shp_readers.hpp
+    */
+
+    void read_polygon_shp(const char *filename,
```

Comment:
This is the only function that should be in `namespace cuspatial`

Reply:
Got it. Will change.

Reply:
read_polygon_shp-->read_polygon_shapefile. Also see comment above.